src/stash/src/upgrade/json_to_proto.rs

```diff
@@ -0,0 +1,1313 @@
+                    .await
+                    .context("migrating storage-shards-to-finalize")?;
+            }
+            collection_name => {
+                // Complain loudly, but don't error out the entire migration. All of the
+                // collections listed above are the ones we know about, and are currently using.
+                // If there exists more it doesn't really matter since nothing in code references
+                // them today, but we want to know if there are.
+                tracing::error!("Unrecognized collection name {collection_name}");
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Migrates a single collection, based on an `id` from the JSON types `(KJ, VJ)` to the protobuf
+/// types `(KP, VP)`.
+async fn migrate_collection<KJ, VJ, KP, VP>(
+    client: &'_ Transaction<'_>,
+    collection_id: i64,
+) -> Result<(), anyhow::Error>
+where
+    KJ: for<'de> Deserialize<'de>,
+    VJ: for<'de> Deserialize<'de>,
+    KP: ::prost::Message + From<KJ>,
+    VP: ::prost::Message + From<VJ>,
+{
+    // Get all of the rows for a single collection and convert them to proto.
+    let mut proto_rows = client
+        .query(
+            "SELECT key, value, time, diff FROM data WHERE collection_id = $1",
+            &[&collection_id],
+        )
+        .await
+        .context("getting rows")?
+        .into_iter()
+        .map(|row| {
+            let key: serde_json::Value = row.get("key");
+            let val: serde_json::Value = row.get("value");
+            let time: i64 = row.get("time");
+            let diff: i64 = row.get("diff");
+
+            let key_json: KJ = serde_json::from_value(key).expect("must deserialize");
+            let val_json: VJ = serde_json::from_value(val).expect("must deserialize");
+
+            let key_proto = KP::from(key_json);
+            let val_proto = VP::from(val_json);
+
+            ((key_proto, val_proto), time, diff)
+        });
+
+    // Insert all of the converted protos into the new table.
+    loop {
+        let mut batch = Vec::new();
+        let mut rows = 0;
+        let mut serialized_size = 0;
+
+        // Keep popping rows into the batch.
+        while let Some(((key, val), time, diff)) = proto_rows.next() {
+            serialized_size += key.encoded_len();
+            serialized_size += val.encoded_len();
+
+            batch.push(((key.encode_to_vec(), val.encode_to_vec()), time, diff));
+            rows += 1;
+
+            // Once our serialized size is greater than our max, break to run an insert.
+            if serialized_size > INSERT_BATCH_MAX_SIZE {
+                break;
+            }
+        }
+
+        // We must have removed all of the rows, exit!
+        if rows == 0 {
+            return Ok(());
+        }
+
+        // Prepare our insert statement.
+        let mut stmt =
+            String::from("INSERT INTO data_proto (collection_id, key, value, time, diff) VALUES");
```

Comment:
Small bit worried for the storage usage collection here which can be hundreds of thousands or millions of rows. Cockroach has a max txn size and I wonder if we'll hit it! Can you do a bit of research to find out what the default max txn size is so we can figure out if we need to worry?

Also concerned about the problem in https://github.com/MaterializeInc/materialize/pull/18916 where there were a few very large rows, even with a tiny row count. This statement will probably succeed in all tests, but fail in prod because of some large user data. Might need to use a similar technique in that PR to do crude byte counting and split up INSERTs.

(Note these are two separate problems with different solutions, and we need to solve both. One is max txn limit. One is max pgwire message size.)

Reply:
I found [this](https://github.com/cockroachdb/docs/blob/master/v1.1/known-limitations.md#write-and-update-limits-for-a-single-transaction), which is fairly old but seems to indicate that the max number of writes (defined as a change to single column, not number of insert statements) is 100,000, with a max possible size of 64MiB. So I think we'll definitely hit that in the case of the storage usage collection. 

Given the way we drop the `data_proto` table before running the migration, and do the swap from `data_proto` to `data` in a separate transaction, I don't think we actually need to run the migration `INSERT`s in a transaction, do we? In which case the only problem we need to worry about is https://github.com/MaterializeInc/materialize/pull/18916 which I believe I already solved by batching inserts.

Reply:
That's so ancient it's absolutely wrong now. Maybe there's no max txn limit anymore? That'd make sense too. So yes, we'd need to batch the rows. But it's not solved because there's no batching going on here?

Reply:
Personally I'd opt to move the inserts outside of a transaction and instead check the epoch before every batch insert. I feel like we'll hit less edge cases this way?

Reply:
That sounds ok as long as this whole migration step can correctly restart if it failed mid way through. Doing the table dropping above sounds like a good way to achieve that?

Reply:
Sounds good! I added a test with failpoints to make sure if we fail between inserting batches or between migrating collections that everything gets properly cleaned up